src/coreclr/jit/valuenum.cpp

```diff
@@ -8857,7 +8885,7 @@ bool Compiler::fgValueNumberConstLoad(GenTreeIndir* tree)
     }
 
     // Throughput check, the logic below is only for USHORT (char)
-    if (!tree->TypeIs(TYP_USHORT))
+    if (!tree->OperIs(GT_IND) || !tree->TypeIs(TYP_USHORT))
```

Comment:
JW, why only `ushort` and not `short` as well? Shouldn't it be the same handling and also cover cases where `short` is used as either an optimization (check for msb set) or done because the underlying instruction set only supports signed types?

Reply:
It's just a TP check, afair the optimization behind it targets string objects, hence, USHORT.